wifi: API to check if an interface is Wi-Fi

krish2718 opened the pull request.

krish2718 committed:
net: l2: Fix the define in comment

This was a typo.

Signed-off-by: Chaitanya Tata <Chaitanya.Tata@nordicsemi.no>

krish2718 committed:
drivers: wifi: Register Wi-Fi architecture type

Identify the Wi-Fi capability to the networking stack and also the type
of Wi-Fi (Native vs Offloaded), this helps identifying Wi-Fi interfaces
that can be used by applications.

Signed-off-by: Chaitanya Tata <Chaitanya.Tata@nordicsemi.no>

krish2718 committed:
net: l2: Add support to identify Wi-Fi interfaces

Wi-Fi is based on L2 Ethernet, so, all drivers are registered as
Ethernet L2, but in order to distinguish true Ethernet and Wi-Fi
devices, add a L2 type within Ethernet.

Also, handle offloaded net devices that also offload Wi-Fi.

This approach is better than adding a new Wi-Fi L2 as that would mean
invasive changes which are unnecessary.

Signed-off-by: Chaitanya Tata <Chaitanya.Tata@nordicsemi.no>

rlubos commented on include/zephyr/net/offloaded_netdev.h:
```diff
@@ -29,6 +29,19 @@ extern "C" {
+	 * Note: To avoid changing drivers, using special value 0 instead
+	 * of BIT(0) for this type.
+	 */
+	L2_OFFLOADED_NET_TYPE_ETHERNET		= 0,
```
0 (default value) should be rather reserved for `UNKNOWN` type, we don't want for example LTE offloaded devices to be represented as Ethernet. Actually, I'm not really sure if we have any offloaded Ethernet devices in the tree.

krish2718 commented on include/zephyr/net/ethernet.h:
```diff
@@ -452,6 +468,9 @@ struct ethernet_api {
 	int (*send)(const struct device *dev, struct net_pkt *pkt);
+
+	/** Types of Ethernet L2 network interfaces */
+	enum ethernet_types l2_eth_type;
```
The reason I have chosen the API is because drivers can easily register the type, putting them in context makes it a tough. 

> Plus, I'm not really convinced we need new enum ethernet_types type - IMO it should be implicit we deal with a Ethernet device, maybe just add bool is_wifi : 1, compiled in only if Wi-Fi is enabled could be added instead?

True, enum is just there for future proofing, esp. I am working on another PR that introduces WIFI_NATIVE, that can be used to handle WPA supplicant (downstream, there is no native Wi-Fi in upstream).

krish2718 replied:
> > The reason I have chosen the API is because drivers can easily register the type, putting them in context makes it a tough.
> 
> That's why I propose to make Ethernet implicit, so that only Wi-Fi devices would need to set the custom flag.

That is what I am doing using default `0` for enum (though a bit convoluted).

> > True, enum is just there for future proofing, esp. I am working on another PR that introduces WIFI_NATIVE, that can be used to handle WPA supplicant (downstream, there is no native Wi-Fi in upstream).
> 
> But is it really different from the device driver perspective, whether the upper layer (supplicant) is offloaded or not?

Well, only driver knows what kind of Wi-Fi functionality is offloaded, so, it gives a hint to the stack and then we can do something like below in Wi-Fi management:

```
if (net_if_wifi_native(iface)) {
   iface->native_api->connect()
else
   off_api->connect()
}
```
And we can have per-iface implementations of Native (I am still thinking about tying native with a iface or keeping it iface agnostic).

rlubos replied:
But such aspects (IMHO) whether to use native or offloaded Wi-Fi management with the specific interface should not be buried inside Ethernet L2, there's a `struct net_wifi_mgmt_offload` (which should probably be renamed at that into something more generic) which could carry whatever Wi-Fi specific configuration you need (likely by grouping native/offloaded APIs into separate structs and setting them accordingly per interface).

I'm also not sure - should native/offloaded Wi-Fi mgmt APIs really differ? IMO it'd be cleaner to have a single mgmt interface (based on the current one) and simply implement it for the native solution. Offloaded interfaces could come up with their own implementations as they do now.

krish2718 commented on include/zephyr/net/net_if.h:
```diff
@@ -2526,6 +2526,67 @@ int net_if_resume(struct net_if *iface);
+ */
+static inline bool net_if_is_wifi(struct net_if *iface)
+{
+	return net_if_is_l2_eth_type_wifi(iface) ||
```
I have now rejigged logic, please check.

krish2718 commented on include/zephyr/net/net_if.h:
```diff
@@ -2526,6 +2526,67 @@ int net_if_resume(struct net_if *iface);
+ */
+static inline bool net_if_is_wifi_offloaded(struct net_if *iface)
+{
+#if defined(CONFIG_NET_OFFLOAD)
```
I have now rejigged logic, please check.

rlubos reviewed (changes requested):
I really don't like putting custom fields into the API structs, no matter how convenient it is.

krish2718 commented:
> I really don't like putting custom fields into the API structs, no matter how convenient it is.

Would this be acceptable? I don't think we even need to rename the struct, as `native == offload_disable`, so, kind of works.

```
$ git diff
diff --git a/include/zephyr/net/wifi_mgmt.h b/include/zephyr/net/wifi_mgmt.h
index 8740d6b4bd..b083d80e33 100644
--- a/include/zephyr/net/wifi_mgmt.h
+++ b/include/zephyr/net/wifi_mgmt.h
@@ -298,6 +298,12 @@ typedef void (*scan_result_cb_t)(struct net_if *iface, int status,
 typedef void (*raw_scan_result_cb_t)(struct net_if *iface, int status,
                                     struct wifi_raw_scan_result *entry);
 #endif /* CONFIG_WIFI_MGMT_RAW_SCAN_RESULTS */
+
+enum wifi_offload_type {
+       /* Native */
+       WIFI_OFFLOAD_DISABLED = 0,
+       WIFI_OFFLOAD_FULL,
+} __packed;
 struct net_wifi_mgmt_offload {
        /**
         * Mandatory to get in first position.
@@ -310,6 +316,7 @@ struct net_wifi_mgmt_offload {
 #else
        struct offloaded_if_api wifi_iface;
 #endif
+       enum wifi_offload_type wifi_off_type;
 
        /* cb parameter is the cb that should be called for each
         * result by the driver. The wifi mgmt part will take care of

```

rlubos commented:
> > I really don't like putting custom fields into the API structs, no matter how convenient it is.
> 
> Would this be acceptable? I don't think we even need to rename the struct, as `native == offload_disable`, so, kind of works.
> 
> ```
> $ git diff
> diff --git a/include/zephyr/net/wifi_mgmt.h b/include/zephyr/net/wifi_mgmt.h
> index 8740d6b4bd..b083d80e33 100644
> --- a/include/zephyr/net/wifi_mgmt.h
> +++ b/include/zephyr/net/wifi_mgmt.h
> @@ -298,6 +298,12 @@ typedef void (*scan_result_cb_t)(struct net_if *iface, int status,
>  typedef void (*raw_scan_result_cb_t)(struct net_if *iface, int status,
>                                      struct wifi_raw_scan_result *entry);
>  #endif /* CONFIG_WIFI_MGMT_RAW_SCAN_RESULTS */
> +
> +enum wifi_offload_type {
> +       /* Native */
> +       WIFI_OFFLOAD_DISABLED = 0,
> +       WIFI_OFFLOAD_FULL,
> +} __packed;
>  struct net_wifi_mgmt_offload {
>         /**
>          * Mandatory to get in first position.
> @@ -310,6 +316,7 @@ struct net_wifi_mgmt_offload {
>  #else
>         struct offloaded_if_api wifi_iface;
>  #endif
> +       enum wifi_offload_type wifi_off_type;
>  
>         /* cb parameter is the cb that should be called for each
>          * result by the driver. The wifi mgmt part will take care of
> ```

Is this really different? IMO we should rather introduce some sort of `get_capabilities()` function at the `struct net_wifi_mgmt_offload` level (but still - net_wifi_mgmt_**offload** for fully native devices simply sounds wrong). Just like generic L2 API struct - it could have `enum net_l2_flags` field directly but it provides a getter function instead, same goes for Ethernet L2 and its capabilities.

krish2718 commented:
> > > I really don't like putting custom fields into the API structs, no matter how convenient it is.
> > 
> > 
> > Would this be acceptable? I don't think we even need to rename the struct, as `native == offload_disable`, so, kind of works.
> > ```
> > $ git diff
> > diff --git a/include/zephyr/net/wifi_mgmt.h b/include/zephyr/net/wifi_mgmt.h
> > index 8740d6b4bd..b083d80e33 100644
> > --- a/include/zephyr/net/wifi_mgmt.h
> > +++ b/include/zephyr/net/wifi_mgmt.h
> > @@ -298,6 +298,12 @@ typedef void (*scan_result_cb_t)(struct net_if *iface, int status,
> >  typedef void (*raw_scan_result_cb_t)(struct net_if *iface, int status,
> >                                      struct wifi_raw_scan_result *entry);
> >  #endif /* CONFIG_WIFI_MGMT_RAW_SCAN_RESULTS */
> > +
> > +enum wifi_offload_type {
> > +       /* Native */
> > +       WIFI_OFFLOAD_DISABLED = 0,
> > +       WIFI_OFFLOAD_FULL,
> > +} __packed;
> >  struct net_wifi_mgmt_offload {
> >         /**
> >          * Mandatory to get in first position.
> > @@ -310,6 +316,7 @@ struct net_wifi_mgmt_offload {
> >  #else
> >         struct offloaded_if_api wifi_iface;
> >  #endif
> > +       enum wifi_offload_type wifi_off_type;
> >  
> >         /* cb parameter is the cb that should be called for each
> >          * result by the driver. The wifi mgmt part will take care of
> > ```
> 
> Is this really different? IMO we should rather introduce some sort of `get_capabilities()` function at the `struct net_wifi_mgmt_offload` level (but still - net_wifi_mgmt_**offload** for fully native devices simply sounds wrong). Just like generic L2 API struct - it could have `enum net_l2_flags` field directly but it provides a getter function instead, same goes for Ethernet L2 and its capabilities.

So, another op similar to `enum ethernet_hw_caps (*get_capabilities)(const struct device *dev);` for Wi-Fi?

rlubos commented:
> So, another op similar to `enum ethernet_hw_caps (*get_capabilities)(const struct device *dev);` for Wi-Fi?

That's what I had in mind. Not sure though, if this should be covered in this PR? It doesn't seem to be strictly related to the original problem this PR is trying to solve (verify if an interface is a Wi-Fi interface).

krish2718 commented:
> > So, another op similar to `enum ethernet_hw_caps (*get_capabilities)(const struct device *dev);` for Wi-Fi?
> 
> That's what I had in mind. Not sure though, if this should be covered in this PR? It doesn't seem to be strictly related to the original problem this PR is trying to solve (verify if an interface is a Wi-Fi interface).

The problem with this is that we need to know if the interface is Wi-Fi before calling `get_capabilities` (As this is not not exposed to net_if, we need to use net_mgmt to get these capabilities). 

Well, whatever way we introduced WIFI_OFFLOAD support in this PR, it will be extended to WIFI_NATIVE and will be used in subsequent PR for wifi_mgmt native handling.

krish2718 commented:
To recap: (Had evaluated many options for this simple flag :) )

We need a way for drivers to tell type of Wi-Fi fundamentally (native, offloaded - default).

* net_if - Too far away for driver, but we could extend `net_if_api` but same concern as this PR (putting flag in API) and also not applicable for all L2s
    - Add a new WIFI_ variants for NET_DEVICE_OFFLOAD_INIT and NET_DEVICE_DT_INST_DEFINE
* L2 - Not possible, as this is driver agnostic
* L2_API -> This PR but a bit ugly putting a flag in the API
* wifi_api -> Not possible, as we don't know if the interface is Wi-Fi capable or not

@rlubos WDYT about #1?

rlubos commented:
> To recap: (Had evaluated many options for this simple flag :) )

So let's try to keep it simple. 

Let's for a moment put aside `wifi_mgmt` offloading and focus on net offloading only.

We have 2 types of devices: native (using Ethernet L2) and offloaded (using offloaded_netdev):
* For native devices - I really can't understand why do you insist on putting the information whether the interface is actually a Wi-Fi into the API structure, and not the Ethernet context. I don't see how:
```
struct ethernet_context *ctx = net_if_l2_data(iface);
```
is more complicated than:
```
const struct ethernet_api *api = net_if_get_device(iface)->api;
```
Or even, if the maintainer has no problem with that, we could simply add a capability whether it's a Wi-Fi device, and use `get_capabilities()` API.

* For offloaded devices - just extend the API structure, but with a function to obtain the device type, not the actual field.


That should be enough to tell whether **any** kind of interface is a Wi-Fi interface or not. And with this information, you can be confident that you work with `struct net_wifi_mgmt_offload`, and put whatever information you need in there (i. e. whether offloaded wifi_mgmt should be used or not).

krish2718 commented:
> > * For offloaded devices - just extend the API structure, but with a function to obtain the device type, not the actual field.
> 
> Ok.

Do we have devices that offload Networking stack but not Wi-Fi? This won't work then, we again need indication from the driver.

krish2718 committed:
net: ethernet: Fix build and doc error

The API should be defined independent of L2_PTP define, this fixes build
error and also the doc error, as the next function will have two doxygen
headers.

Signed-off-by: Chaitanya Tata <Chaitanya.Tata@nordicsemi.no>

rlubos commented on include/zephyr/net/ethernet.h:
```diff
@@ -347,6 +347,19 @@ enum ethernet_filter_type {
 
+/** Types of Ethernet L2 */
+enum ethernet_if_types {
+	/** IEEE 802.3 Ethernet */
+	L2_ETH_IF_TYPE_ETHERNET			= BIT(0),
```
Although not likely to ever be used - wouldn't defining type like this (i. .e bitfields) require changes in **all** existing Ethernet interfaces? Otherwise they end up with unspecified `eth_if_type`. I still think though, that Ethernet should be implicit and Wi-Fi should just get a flag (or capability field).

krish2718 replied:
I have now removed bit fields, so, left the enum with a comment that this is default.

rlubos commented on include/zephyr/net/ethernet.h:
```diff
@@ -347,6 +347,19 @@ enum ethernet_filter_type {
+	/** IEEE 802.11 Wi-Fi*/
+	L2_ETH_IF_TYPE_WIFI			= BIT(1),
+
+	/** IEEE 802.11 Wi-Fi Offloaded (0) or Native (1) */
+	L2_ETH_IF_TYPE_WIFI_NATIVE		= BIT(2),
```
This is IMO ambiguous - merely the fact that the interface uses Ethernet L2 implies it's native.

krish2718 replied:
There could be a Wi-Fi chip that deals with Ethernet frames but Wi-Fi is fully offloaded, no?

rlubos commented on include/zephyr/net/offloaded_netdev.h:
```diff
@@ -29,6 +29,18 @@ extern "C" {
 
+/** Types of offloaded netdev L2 */
+enum offloaded_net_if_types {
+	/** Ethernet devices */
+	L2_OFFLOADED_NET_IF_TYPE_ETHERNET	= BIT(0),
+
+	/** Modem */
+	L2_OFFLOADED_NET_IF_TYPE_MODEM		= BIT(1),
+
+	/** IEEE 802.11 Wi-Fi */
+	L2_OFFLOADED_NET_IF_TYPE_WIFI		= BIT(2),
```
Shouldn't the device types be mutually exclusive? Not sure I understand why is this defined as a bitfield.

krish2718 replied:
Yeah, I was implementing them differently and this was a leftover, will clean it up.

rlubos commented:
I'll be off for the rest of the week, please forgive delayed responses.

krish2718 commented on include/zephyr/net/net_if.h:
```diff
@@ -2526,6 +2526,24 @@ int net_if_resume(struct net_if *iface);
+bool net_if_is_wifi(struct net_if *iface);
+
+/**
+ * @brief Check if the network interface used Native Wi-Fi.
```
In upstream this isn't applicable, but for downstream Native Wi-Fi would mean not using the `off_api` e.g., WPA supplicant implementing Wi-Fi management ops.

krish2718 replied:
Thinking a bit futuristic, there might be multiple Native solutions e.g., WPA supplicant, iwd etc. And each driver can choose its own and they should be able to coexist wlan0-WPA supplicant, wlan1-iwd etc.

krish2718 replied:
This is only about identifying if a iface is Wi-Fi or not, so, the API/Flag has to be outside Wi-Fi API, no? So, I have added to any L2 that supports Wi-Fi (Ethernet and Offloaded Net device).

rlubos replied:
So if it's **only** about identifying whether an interface is a Wi-Fi or not, then I really don't understand why are there two identifiers for Wi-Fi device type in Ethernet L2 (`L2_ETH_IF_TYPE_WIFI` vs `L2_ETH_IF_TYPE_WIFI_NATIVE`). `L2_ETH_IF_TYPE_WIFI` should be enough to tell that it's safe to use Wi-Fi APIs with the interface.

krish2718 replied:
Those are also needed to decide which API to invoke (`off_api` or `native_api`), so, also cannot be part of the Wi-Fi API.

krish2718 committed:
net: wifi: Add support to get and set Wi-Fi as default interface

Add a configuration option to set Wi-Fi as default interface and also
add an API to get first available Wi-Fi interface.

Signed-off-by: Chaitanya Tata <Chaitanya.Tata@nordicsemi.no>

krish2718 committed:
net: wifi: Use only Wi-Fi interfaces

As this is Wi-Fi shell use only Wi-Fi interfaces, if none are found
fail.

Signed-off-by: Chaitanya Tata <Chaitanya.Tata@nordicsemi.no>

krish2718 commented:
Ping! Another PR is depending on this, so, good to get this merged to avoid rebase hell :).

@sylvioalves can I trouble you for review on ESP parts in this and the other PR?

dleach02 commented on include/zephyr/net/net_if.h:
```diff
@@ -2526,6 +2526,23 @@ int net_if_resume(struct net_if *iface);
+ *
+ * @return Pointer to network interface, NULL if not found.
+ */
+struct net_if *net_if_get_first_wifi(void);
```
question, why the first wifi IF? If the platform has <n> wifi interfaces how would one specify an arbitrary IF?

krish2718 replied:
This is just a commonly used helper, one can do `net_if_foreach` + `net_if_is_wifi` and choose any Wi-Fi interface of interest based on any criterion. If its needed we can even add `net_if_wifi_foreach` in the future as a shortcut.

mniestroj commented on subsys/net/ip/net_if.c:
```diff
@@ -4592,6 +4593,27 @@ void net_if_add_tx_timestamp(struct net_pkt *pkt)
+		return net_off_is_wifi_offloaded(iface);
+	}
+#if defined(CONFIG_NET_L2_ETHERNET)
+	return net_eth_type_is_wifi(iface);
```
This comment was addressed, but in the next commit (not the one that introduced `net_eth_type_is_wifi(iface)`) call.
